Feature/create context for view mode

OlegKotsubko opened the pull request:
## Added app state management for Edit  and Preview modes

## Checklist
- ViewModeContextProvider code base
- delete Note while editing
- delete Note while preview mode

OlegKotsubko committed:
Create ViewModeContext and implement methods onEdit and onPreview.
Need to refactor unnecessary code

OlegKotsubko committed:
Added new functional.
- can delete Note when edit mode is enabled
- can delete Note when preview mode is enabled

Removed unnecessary methods from NotesContext

OlegKotsubko committed:
Added view mode context

OlegKotsubko committed:
rem typescript

OlegKotsubko committed:
revert yarn lock

Rastopyr commented on src/App.js:
```diff
@@ -1,49 +1,38 @@
-  } = useViewModeContext()
+  const LeftSidebar = () => {
+    switch (mode) {
+    case "PREVIEW":
```
Тут лучше использовать не строковые литералы, а константы или вообще описать возможные режимы объектом. Это проще поддерживать и менять.

\+ если будет система типов(flow, typescript) то switch будет еще типизированный

Rastopyr commented on src/App.js:
```diff
@@ -1,49 +1,38 @@
+        )
+    default:
+      return (
+        <FormContainer
```
Давай для формы пропишем отдельный мод явно?

Rastopyr commented on src/actions/view-actions.js:
```diff
@@ -0,0 +1,18 @@
+export const SET_PREVIEW = "SET_PREVIEW"
```
`SET_PREVIEW` -> `SET_PREVIEW_ROUTE_MODE`
`SET_EDIT` -> `SET_EDIT_ROUTE_MODE`
`SET_DEFAULT` -> `SET_DEFAULT_ROUTE_MODE`

Rastopyr commented on src/actions/view-actions.js:
```diff
@@ -0,0 +1,18 @@
+export const SET_EDIT = "SET_EDIT"
+export const SET_DEFAULT = "SET_DEFAULT"
+
+export const setPreview = (dispatch) => () =>
```
`setPreview` -> `setPreviewMode`
`setEdit` -> `setEditMode`
`setDefault` -> `setDefaultMode`

Rastopyr commented on src/components/FormContainer/FormContainer.js:
```diff
@@ -0,0 +1,37 @@
+  switch (mode) {
+    case "EDIT":
+      return (
+        <Form
```
Тут лучше использовать не строковые литералы, а константы или вообще описать возможные режимы объектом. Это проще поддерживать и менять.

\+ если будет система типов(flow, typescript) то switch будет еще типизированный

Rastopyr commented on src/contexts/ViewModeContext.js:
```diff
@@ -1,28 +1,29 @@
-    setPreview(!isPreview)
-  }
+  const [modeActions] = useState(() => ({
+    onPreview: setPreview(dispatch),
```
Именование `onWhatever` лучше использовать когда речь идет про событие. В данном случае речь идет про API. Тут лучше подойдет императивное именование `setPreviewMode` или `goToPreview`

Rastopyr commented on src/contexts/ViewModeContext.js:
```diff
@@ -1,28 +1,29 @@
-      toggleEdit,
-      togglePreview,
+      mode,
+      modeActions,
```
`modeActions` -> `actions`

OlegKotsubko replied:
<img width="239" alt="Screenshot 2022-02-15 at 20 06 07" src="https://user-images.githubusercontent.com/18003430/154122333-d8ab3368-1cab-44fa-8138-58b02cd85045.png">
если использовать два хука рядом будет конфликт имен или придется писать `actions: modeActions`

Rastopyr replied:
> если использовать два хука рядом будет конфликт имен или придется писать `actions: modeActions` 

Да, консьюмер уже должен соответствующим образом называть поля и методы

OlegKotsubko committed:
decreased coupling
- removed dependencies from Components
- rewrote actions for view mode
- rewrote view reducer and hook
- added default mode state in App

Rastopyr commented on src/reducer/view-reducer.js:
```diff
@@ -1,17 +1,23 @@
+  SET_EDIT_ROUTE_MODE,
 } from "../actions/view-actions";
 
+export const MODE = {
```
`MODE` -> `VIEW_MODE`

Rastopyr commented on src/components/NotesList/NotesList.js:
```diff
@@ -1,29 +1,40 @@
+          id={id}
+          title={title}
+          description={rawTextFromHTML}
+          notesActions={notesActions}
```
Зачем сюда передается notesActions?

Rastopyr replied:
Тоже самое про viewModelActions

OlegKotsubko committed:
Rename modeActions to actions

Rastopyr commented on src/components/NotesList/NotesList.js:
```diff
@@ -1,29 +1,40 @@
+          notesActions={notesActions}
+          viewModeActions={viewModeActions}
+          setActiveNoteById={setActiveNoteById}
+          deleteActiveNote={deleteActiveNote}
```
`onDelete`

Rastopyr commented on src/components/NotesList/NotesList.js:
```diff
@@ -1,29 +1,40 @@
+          description={rawTextFromHTML}
+          notesActions={notesActions}
+          viewModeActions={viewModeActions}
+          setActiveNoteById={setActiveNoteById}
```
`onActivate`

Rastopyr commented on src/components/NotesList/NotesList.js:
```diff
@@ -1,29 +1,40 @@
+          viewModeActions={viewModeActions}
+          setActiveNoteById={setActiveNoteById}
+          deleteActiveNote={deleteActiveNote}
+          noteRemoveHandler={noteRemoveHandler}
```
onRemove

OlegKotsubko committed:
Rewrite single note from overprops

OlegKotsubko committed:
Add on reset activation handler

OlegKotsubko committed:
Separate methods for NoteList to NoteListView file

OlegKotsubko committed:
Added flag to edit mode in Form
